File: sdk/typescript/src/device.ts
import { Transport } from './transport.js';
import { DeviceStats } from './types.js';

export class Device {
  private transport: Transport;

  constructor(transport: Transport) {
    this.transport = transport;
  }

  async stats(): Promise<DeviceStats> {
    try {
      return await this.transport.getJson<DeviceStats>('/api/device/stats');
    } catch {
      const isHealthy = await this.health();
      return {
        status: isHealthy ? 'running' : 'unreachable',
        port: 8080,
      };
    }
  }

  async health(): Promise<boolean> {
    try {
      const res = await this.transport.getJson<{ status?: string }>('/health');
      return res?.status === 'ok' || res?.status === 'loading model';
    } catch {
      return false;
    }
  }
}

File: sdk/python/buildify/device.py
"""Device telemetry, status, and health client."""

from typing import Dict, Any, Optional
from .types import DeviceStats
from .transport import BuildifyTransport, AsyncBuildifyTransport


class Device:
    """Synchronous device telemetry and management."""

    def __init__(self, transport: BuildifyTransport):
        self._transport = transport

    def stats(self) -> DeviceStats:
        """Retrieves phone battery, thermal, and server status."""
        try:
            data = self._transport.get_json("/api/device/stats")
            return DeviceStats.model_validate(data)
        except Exception:
            # Fallback when only llama-server is active without full backend daemon
            health = self.health()
            return DeviceStats(status="running" if health else "unreachable")

    def health(self) -> bool:
        """Checks if the edge inference server is responding."""
        try:
            data = self._transport.get_json("/health")
            return data.get("status") in ("ok", "loading model")
        except Exception:
            return False


class AsyncDevice:
    """Asynchronous device telemetry and management."""

    def __init__(self, transport: AsyncBuildifyTransport):
        self._transport = transport

    async def stats(self) -> DeviceStats:
        """Retrieves phone battery, thermal, and server status."""
        try:
            data = await self._transport.get_json("/api/device/stats")
            return DeviceStats.model_validate(data)
        except Exception:
            health = await self.health()
            return DeviceStats(status="running" if health else "unreachable")

    async def health(self) -> bool:
        """Checks if the edge inference server is responding."""
        try:
            data = await self._transport.get_json("/health")
            return data.get("status") in ("ok", "loading model")
        except Exception:
            return False
